olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -155,18 +153,177 @@ client::CancellableFuture<DataResponse> VersionedLayerClientImpl::GetData(
+          auto handle = it->second;
+          auto context_it = contexts.emplace(contexts.end());
+
+          AddTask(task_scheduler, pending_requests,
+                  [=](CancellationContext inner_context) {
+                    // Get blob data
+                    auto data = repository::DataRepository::GetVersionedData(
+                        catalog, layer_id,
+                        DataRequest().WithDataHandle(handle).WithBillingTag(
+                            request.GetBillingTag()),
+                        inner_context, settings);
+
+                    if (!data.IsSuccessful()) {
+                      promise->set_value(std::make_shared<PrefetchTileResult>(
+                          tile, data.GetError()));
+                    } else {
+                      promise->set_value(std::make_shared<PrefetchTileResult>(
+                          tile, PrefetchTileNoError()));
+                    }
+
+                    flag->exchange(true);
+                    return EmptyResponse(PrefetchTileNoError());
+                  },
+                  [=](EmptyResponse result) {
+                    if (!flag->load()) {
+                      // If above task was cancelled we might need to set
+                      // promise else below task will wait forever
+                      promise->set_value(std::make_shared<PrefetchTileResult>(
+                          tile,
+                          client::ApiError(ErrorCode::Cancelled, "Cancelled")));
+                    }
+                  },
+                  *context_it);
+          it++;
+        }
+
+        // Task to wait for previously triggered data download to collect
+        // responses and trigger user callback.
+        AddTask(
+            task_scheduler, pending_requests,
+            [=](CancellationContext inner_context) -> PrefetchTilesResponse {
+              PrefetchTilesResult result;
+              result.reserve(futures->size());
+
+              for (auto& future : *futures) {
+                // Check if cancelled in between.
+                if (inner_context.IsCancelled()) {
+                  return {{ErrorCode::Cancelled, "Cancelled"}};
+                }
+
+                auto tile_result = future.get();
+                result.emplace_back(std::move(tile_result));
+              }
+
+              OLP_SDK_LOG_INFO_F(kLogTag, "Prefetch done, key=%s, tiles=%zu",
+                                 key.c_str(), result.size());
+              return PrefetchTilesResponse(std::move(result));
+            },
+            callback, *contexts.emplace(contexts.end()));
+
+        context.ExecuteOrCancelled([&]() {
+          return client::CancellationToken([contexts]() {
+            for (auto context : contexts) {
+              context.CancelOperation();
+            }
+          });
+        });
+
+        return EmptyResponse(PrefetchTileNoError());
+      },
+      // No callback as we only return status in case of error, cancellation
```

Comment:
I would specify the comment more explicitly, smth like that:
'Because the handling of prefetch tiles responses is performed by the inner-task, no need to set a callback here. Otherwise, the user would be notified with empty results.'